CHAID/CHAID.py

```diff
@@ -366,46 +377,85 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
         for i, index in enumerate(ind):
             index = index.deep_copy()
             unique = set(index.arr)
 
             mappings = MappingDict()
-            frequencies = {}
+            freq = {}
+            wt_freq = {}
             for col in unique:
                 counts = np.unique(dep.arr[index.arr == col], return_counts=True)
-                frequencies[col] = cl.defaultdict(int)
-                frequencies[col].update(np.transpose(counts))
+                freq[col] = cl.defaultdict(int)
+                freq[col].update(np.transpose(counts))
+                if wt is not None:
+                    wt_freq[col] = cl.defaultdict(int)
+                    for dep_v in set(dep.arr):
+                        wt_freq[col][dep_v] = wt[(index.arr == col) * (dep.arr == dep_v)].sum()
 
             while len(unique) > 1:
                 size = int((len(unique) * (len(unique) - 1)) / 2)
-                sub_data_columns = [('combinations', object), ('chi', float), ('p', float)]
+                sub_data_columns = [('combinations', object), ('p', float), ('chi', float)]
                 sub_data = np.array([(None, 0, 1)]*size, dtype=sub_data_columns, order='F')
                 for j, comb in enumerate(it.combinations(unique, 2)):
-                    col1_freq = frequencies[comb[0]]
-                    col2_freq = frequencies[comb[1]]
+                    col1_freq = freq[comb[0]]
+                    col2_freq = freq[comb[1]]
 
                     keys = set(col1_freq.keys()).union(col2_freq.keys())
 
-                    cr_table = [
+                    n_ij = np.array([
                         [col1_freq.get(k, 0) for k in keys],
                         [col2_freq.get(k, 0) for k in keys]
-                    ]
-
-                    chi = stats.chi2_contingency(np.array(cr_table), correction=False)
-                    sub_data[j] = (comb, chi[0], chi[1])
+                    ])
+
+                    if wt is None:
+                        m_ij = (np.vstack(n_ij.sum(axis=1)) * n_ij.sum(axis=0)) / n_ij.sum().astype(float)
+                    else:
+                        col1_wt_freq = wt_freq[comb[0]]
+                        col2_wt_freq = wt_freq[comb[1]]
+
+                        m_ij = n_ij / np.array([
+                            [col1_wt_freq.get(k, 0) for k in keys],
+                            [col2_wt_freq.get(k, 0) for k in keys]
+                        ])
+                        m_ij = self.weighted_case(n_ij, m_ij)
+
+                    dof = (n_ij.shape[0] - 1) * (n_ij.shape[1] - 1)
+                    ret = st_chi(n_ij, f_exp=m_ij, ddof=n_ij.size - 1 - dof, axis=None)
+
+                    sub_data[j] = (comb, ret[1], ret[0])
+
+                choice, highest_p_join, chi_join = max(sub_data, key=lambda x: (x[1], x[2]))
+
+                if highest_p_join < self.alpha_merge:
+                    if len(freq.keys()) == 2 or len(wt_freq.keys()) == 2: # have actual p-value
```

Comment:
I'm not sure this branch really adds much in the way of speed, and it makes the code harder to follow. I'd remove it.